src/kernel/src/processor.rs

```diff
@@ -355,3 +381,142 @@ pub fn register(id: u32, bsp_id: u32) {
         }
     }
 }
+
+fn enqueue_ipi_task_many(incl_self: bool, task: &Arc<IpiTask>) {
+    let current = current_processor();
+    for p in unsafe { &ALL_PROCESSORS }.iter().flatten() {
+        if p.id != current.id || incl_self {
+            p.enqueue_ipi_task(task.clone());
+        }
+    }
+}
+
+/// Run a closure on some set of CPUs, waiting for all invocations to complete.
+pub fn ipi_exec(target: Destination, f: Box<dyn Fn() + Send + Sync>) {
+    let task = Arc::new(IpiTask {
+        outstanding: AtomicU64::new(0),
+        func: f,
+    });
+
+    // We need to disable interrupts to prevent our current CPU from changing until we've submitted the IPIs.
+    let int_state = interrupt::disable();
+    let current = current_processor();
+    match target {
+        // Lowest priority doesn't really make sense in IPIs, so we just pretend it goes to BSP.
+        Destination::Bsp | Destination::LowestPriority => {
+            get_processor(current.bsp_id()).enqueue_ipi_task(task.clone());
+        }
+        Destination::Single(id) => {
+            let proc = get_processor(id);
+            if !proc.is_running() {
+                logln!("tried to send IPI to non-running CPU");
+                interrupt::set(int_state);
+                return;
+            }
+            if proc.id == current.id {
+                // We are the only recipients, so just run the closure.
+                (task.func)();
+                interrupt::set(int_state);
+                return;
+            }
+            proc.enqueue_ipi_task(task.clone());
+        }
+        Destination::AllButSelf => enqueue_ipi_task_many(false, &task),
+        Destination::All => enqueue_ipi_task_many(true, &task),
+    }
+
+    // No point using the IPI hardware to send ourselves a message, so just run it manually if current CPU is included.
+    let (target, target_self) = match target {
+        Destination::All => (Destination::AllButSelf, true),
+        x => (x, false),
+    };
+    arch::send_ipi(target, GENERIC_IPI_VECTOR);

```

Comment:
Do we still have to do this if it is on the same CPU?

Reply:
Do you mean if target is Destination::All?

Reply:
Ok I see, looks fine.